### PR TITLE
hw7 ml Шарова Татьяна

### PR DESCRIPTION
**Pull request by Verbarius**

### CODE REVIEW

**Comment by Verbarius** [comment]
from sklearn.datasets import load_iris
data = load_iris()
x = data['data']
y = data['target']
xtrain, xtest, ytrain, ytest = train_test_split(x, y, test_size=0.2, random_state=1)
lr1 = linear_model.LogisticRegression()
lr2 = Multiclass(lambda_coef = 0.003)
lr1.fit(xtrain, ytrain)
lr2.multi_fit(xtrain, ytrain)
loss1 = log_loss(ytest, lr1.predict_proba(xtest))
loss2 = log_loss(ytest, lr2.predict_proba(xtest))
print('my_loss: ', loss2, ' vs loss: ', loss1)

OUTPUT: my_loss:  0.6539610196501595  vs loss:  0.3909330389867508

**Comment by krabye** [comment]
Homework 7 ml soft deadline.

**Comment by krabye** [comment]
Hard deadline hw7.

File: homeworks/homework_07_ml/logistic_regression.py [hunk 1]
```diff
@@ -23,27 +31,116 @@ def fit(self, X_train, y_train):
...
+            self.w += self.lambda_coef * grad/self.batch_size
+            err = self.log_loss(x, y, self.w)/self.batch_size
+            self.loss.append(err)
+
+    def sigmoid(self, z):
+        return 1/(1 + np.exp(-np.clip(z, -250, 250)))
+
+    def _grad(self, x, y, w):
+        gr = np.empty_like(w)
+        for j in range(w.shape[1]):
```
Comment (krabye) [inline homeworks/homework_07_ml/logistic_regression.py:52]: Матричные операции должны быть без циклов.

File: homeworks/homework_07_ml/logistic_regression.py [hunk 2]
```diff
@@ -23,27 +31,116 @@ def fit(self, X_train, y_train):
...
+            gr[0, j] = np.sum((y.reshape(-1, 1) - self.sigmoid(x.dot(w.T)).reshape(-1, 1))*x[:, j])
+            if self.regularization == 'elastic_net':
```
Comment (krabye) [inline homeworks/homework_07_ml/logistic_regression.py:54]: А если бы 'l2' передал, то просто не было бы регуляризации? :)

File: homeworks/homework_07_ml/logistic_regression.py [hunk 3]
```diff
@@ -23,27 +31,116 @@ def fit(self, X_train, y_train):
...
+        n = X_train.shape[1]
+        y_train = y_train.reshape((-1, 1))
+        table = np.hstack((np.ones((X_train.shape[0], 1)), X_train, y_train))
+        self.w = np.random.normal(scale=1e-5, size=(1, n+1))
+        for iter_ in range(self.max_iter):
+            perm = np.random.permutation(table)[:self.batch_size]
+            x = perm[:, :-1]
+            y = perm[:, -1].reshape(-1, 1)
+            grad = self._grad(x, y, self.w)
+            self.w += self.lambda_coef * grad/self.batch_size
```
Comment (krabye) [inline homeworks/homework_07_ml/logistic_regression.py:43]: Приехали, у нас градиентный спуск или градиентный подъем?

Comment (krabye) [reply homeworks/homework_07_ml/logistic_regression.py:43]: Понимаю, что в формуле градиента ты знак минуса внесла за скобки, однако в регуляризации ты про знак забыла.

Comment (krabye) [reply homeworks/homework_07_ml/logistic_regression.py:43]: И деление на размер батча у нас фигурирует только в логлосе, никакого размера батча в регуляризации нет.

File: homeworks/homework_07_ml/logistic_regression.py [hunk 4]
```diff
@@ -23,27 +31,116 @@ def fit(self, X_train, y_train):
...
+    def __init__(self, lambda_coef=0.001, beta=0.5, alpha=0.5, batch_size=50, max_iter=1000):
+        super().__init__()
+        self.w_matrix = 0
+        self.lambda_coef = lambda_coef
+        self.alpha = alpha
+        self.beta = beta
+        self.batch_size = batch_size
+        self.max_iter = max_iter
+
+    def multi_fit(self, X_train, y_train):
```
Comment (krabye) [inline homeworks/homework_07_ml/logistic_regression.py:104]: Сори, это не софтмакс, а one-vs-rest.
